[vcpkg registries] support versions

strega-nil opened the pull request:
This PR merges the Registries changes and the versioning changes, so that one can use both at the same time.

There is one major difference between this PR and the RFC (#13590), which is that instead of version files looking like:

```json
[
  ...
]
```

version files look like:

```
{
  "versions": [
    ...
  ]
}
```

this is to support interop between this PR and existing demos and the like;
fixing this, along with perhaps renaming `port_versions` to `port-versions` should be done after this is merged,
should be a trivial change.

This is the first PR resulting from splitting #15054 into multiple PRs.

The e2e tests for filesystem registries will come in the git registries PR, since there's already test code written for that.

strega-nil committed:
[vcpkg registries] support versions

This PR merges the Registries changes and the versioning changes, so that one can use both at the same time.

There is one major difference between this PR and the RFC (#13590), which is that instead of version files looking like:

```json
[
  ...
]
```

version files look like:

```
{
  "versions": [
    ...
  ]
}
```

this is to support interop between this PR and existing demos and the like;
fixing this, along with perhaps renaming `port_versions` to `port-versions` should be done after this is merged,
should be a trivial change.

strega-nil committed:
robert CRs

strega-nil commented on toolsrc/src/vcpkg/base/files.cpp:
```diff
@@ -1192,7 +1192,7 @@ namespace vcpkg::Files
             }
 
-#if defined(WIN32)
+#if defined(_WIN32)
```
drive-by fix.

strega-nil commented on toolsrc/src/vcpkg/base/files.cpp:
```diff
@@ -1228,7 +1228,7 @@ namespace vcpkg::Files
                 if (fd == -1)
                 {
-                    fd = ::open(native.c_str(), 0);
+                    fd = ::open(native.c_str(), O_RDWR | O_CREAT, S_IRUSR | S_IWUSR | S_IRGRP | S_IROTH);
```
This allows `::open` to create a file that doesn't already exist.

strega-nil commented on toolsrc/src/vcpkg/versiondeserializers.cpp:
```diff
@@ -21,6 +21,7 @@ namespace
 
         Optional<std::string> visit_string(Json::Reader& r, StringView sv) override
         {
+            (void)GIT_TREE;
```
```suggestion
```

strega-nil commented on toolsrc/src/vcpkg/vcpkgpaths.cpp:
```diff
@@ -222,7 +233,12 @@ namespace vcpkg
             FeatureFlagSettings m_ff_settings;
+
+            fs::path registries_work_tree_dir() const { return m_registries_path / fs::u8path("git"); }
+            fs::path registries_dot_git_dir() const { return registries_work_tree_dir() / fs::u8path(".git"); }
```
These lines should be removed.

strega-nil commented on toolsrc/src/vcpkg/registries.cpp:
```diff
@@ -16,350 +18,715 @@ namespace
             RegistryImplDeserializer::KIND,
+            RegistryImplDeserializer::BASELINE,
             RegistryImplDeserializer::PATH,
+            RegistryImplDeserializer::REPO,
```
Should be deleted.

strega-nil reviewed (commented):
These are the things I noticed and am fixing now, along with some notes.

strega-nil committed:
Fix nicole CRs.

strega-nil committed:
fix minor logic bug in overlay ports

BillyONeal commented on toolsrc/src/vcpkg/portfileprovider.cpp:
```diff
@@ -95,21 +86,25 @@ namespace vcpkg::PortFileProvider
+                    if (scf->core_paragraph->name == spec)
                     {
-                        return SourceControlFileLocation{std::move(*scf), ports_dir};
+                        auto res = std::make_unique<OverlayRegistryEntry>();
```
This somewhat suggests a missing ctor of `OverlayRegistryEntry` since it's doing the usual "create + assign over" antipattern.

BillyONeal replied:
Ditto below.

BillyONeal commented on toolsrc/include/vcpkg/registries.h:
```diff
@@ -17,40 +18,46 @@ namespace vcpkg
         // always ordered lexicographically
         View<std::string> packages() const { return packages_; }
-        const RegistryImpl& implementation() const { return *implementation_; }
+        const RegistryImplementation& implementation() const { return *implementation_; }
```
I think it would be a better design for `Registry` to replicate any function call that someone would call `implementation()` over for law-of-demeter reasons.

(That said this is preexisting in this PR so no change necessarily requested)

BillyONeal commented on toolsrc/src/vcpkg/registries.cpp:
```diff
@@ -16,341 +18,693 @@ namespace
-                auto it = res.versions.lower_bound(buffer.first);
-                if (it == res.versions.end() || it->first != buffer.first)
+                auto filename = fs::u8string(file.path().filename());
+                if (Strings::ends_with(filename, ".json"))
```
Consider `if (!condition) { continue; }` to reduce extreme indenting (like 169)

strega-nil committed:
add a comment explaining GitRegistryEntry

BillyONeal commented on toolsrc/src/vcpkg/registries.cpp:
```diff
@@ -16,341 +18,693 @@ namespace
+            }
             return res;
         }
+        else
```
Delete else due to preceeding return?

BillyONeal commented on toolsrc/src/vcpkg/registries.cpp:
```diff
@@ -16,341 +18,693 @@ namespace
+                {
+                    auto& scf = *scfp;
+                    auto maybe_error = scf->check_against_feature_flags(port_directory, paths.get_feature_flags());
+                    if (maybe_error)
```
Add {}s

BillyONeal commented on toolsrc/src/vcpkg/registries.cpp:
```diff
@@ -16,341 +18,693 @@ namespace
-                    return std::make_unique<FilesystemEntry>(std::move(*pentry));
+                    return std::move(*p);
                 }
                 else
```
Delete else due to preceeding return?

BillyONeal commented on toolsrc/src/vcpkg/registries.cpp:
```diff
@@ -16,341 +18,693 @@ namespace
+
+            // attempt to check out the baseline:
+            auto maybe_path = get_git_baseline_json_path(paths, m_baseline_identifier);
+            if (auto path_to_git_baseline = maybe_path.get())
```
Maybe flip this and delete the else for less indents?

BillyONeal commented on toolsrc/src/vcpkg/registries.cpp:
```diff
@@ -16,341 +18,693 @@ namespace
-            const auto& fs = paths.get_filesystem();
-            auto entry_path = this->path_to_port_entry(paths, port_name);
-            if (!fs.exists(entry_path))
+        const auto& baseline = m_baseline.get([this, &paths]() -> Baseline {
```
I'm really not a fan of lambdas that are so long they don't fit on a screen because it makes figuring out what `return` means therein somewhat obscure. Can this be made a named member `const Baseline& BuiltinRegistry::populate_baseline(const VcpkgPaths&)` or similar?

BillyONeal commented on toolsrc/src/vcpkg/registries.cpp:
```diff
@@ -16,341 +18,693 @@ namespace
+        {
+            return it->second;
+        }
+        else
```
Delete else due to included return?

BillyONeal commented on toolsrc/src/vcpkg/registries.cpp:
```diff
@@ -375,9 +729,163 @@ namespace vcpkg
+        }
+
+        auto maybe_contents = fs.read_contents(versions_file_path);
+        if (auto contents = maybe_contents.get())
```
Flip and delete else for -1 indent?

BillyONeal reviewed (commented):
I feel bad that I don't have substantial code review comments, because most of this is parsing code etc. and I don't think humans in general are very good at reviewing that without debugging through examples or having complete test coverage.

I do think this needs some tests before it is merged. (I don't if versions had test coverage but had I been around to review that I'd probably have made the same comment :/ )

strega-nil commented on toolsrc/include/vcpkg/base/files.h:
```diff
@@ -236,6 +236,7 @@ namespace vcpkg::Files
         void current_path(const fs::path& path, LineInfo li);
 
         // waits forever for the file lock
+        // creates a non-existent file
```
It is attempting to say "if you pass a path that doesn't exist, this function creates that file and locks it."

BillyONeal replied:
See I interpreted this as "it is an error if the file exists" because "non-existent" file looks like a precondition.

strega-nil replied:
mmh, I'll reword the comment.

strega-nil commented on toolsrc/include/vcpkg/base/jsonreader.h:
```diff
@@ -75,13 +75,13 @@ namespace vcpkg::Json
         };
         std::vector<Path> m_path;
 
+    public:
```
Probably, but this is a small change to just allow a single additional public member.

strega-nil commented on toolsrc/src/vcpkg/install.cpp:
```diff
@@ -839,17 +841,12 @@ namespace vcpkg::Install
+                auto baseprovider = std::make_unique<PortFileProvider::BaselineProvider>(paths);
+                if (auto p_baseline = manifest_scf.core_paragraph->extra_info.get("$x-default-baseline"))
+                {
+                    paths.get_configuration().registry_set.experimental_set_builtin_registry_baseline(
```
Yes. This function will be removed, and it's important to say so, I think.

strega-nil committed:
Merge remote-tracking branch 'upstream/master' into registries-versions-merge

strega-nil committed:
Billy CRs

strega-nil committed:
add unit tests for RegistrySet

strega-nil committed:
testing: add e2e testing for registries

strega-nil committed:
Add e2e tests

found a bug!

ras0219-msft reviewed (commented):
Review still in progress (portfileprovider.cpp:299)

strega-nil commented on toolsrc/include/vcpkg/configuration.h:
```diff
@@ -13,4 +17,7 @@ namespace vcpkg
     };
+
+    std::unique_ptr<Json::IDeserializer<Configuration>> make_configuration_deserializer(
+        const VcpkgCmdArguments& args, const fs::path& config_directory);
```
This is extremely hard to actually implement.

strega-nil commented on toolsrc/include/vcpkg/registries.h:
```diff
@@ -17,40 +19,46 @@ namespace vcpkg
-        virtual fs::path get_port_directory(const VcpkgPaths& paths, const VersionT& version) const = 0;
+        virtual View<VersionT> get_port_versions() const = 0;
+
+        // precondition: version \in get_port_versions(). Otherwise, this function asserts.
```
This is actually how it works in the implementation, fwiw.

strega-nil committed:
Robert CRs, e2e test fixing

switch to using the same git archive trick which is used in the git registry pr

strega-nil committed:
format

strega-nil committed:
add --no-hardlinks

strega-nil committed:
Switch back to old way of doing check out

BillyONeal reviewed (approved):
I still would really like to see unit level tests for any parsing code that aren't present here.

strega-nil commented on toolsrc/src/vcpkg-test/registries.cpp:
```diff
@@ -0,0 +1,69 @@
+        {
+            return tri->number;
+        }
+        else
+        {
+            return -1;
+        }
```
Given that this is basically a pattern match, not a lot of code, not going to use this suggestion.

strega-nil committed:
billy crs

strega-nil committed:
format

strega-nil committed:
add registry parsing unit tests

strega-nil commented:
/azp run

strega-nil commented on toolsrc/include/vcpkg/vcpkgcmdarguments.h:
```diff
@@ -134,6 +134,8 @@ namespace vcpkg
         std::unique_ptr<std::string> scripts_root_dir;
         constexpr static StringLiteral BUILTIN_PORTS_ROOT_DIR_ARG = "x-builtin-ports-root";
         std::unique_ptr<std::string> builtin_ports_root_dir;
+        constexpr static StringLiteral BUILTIN_PORT_VERSIONS_DIR_ARG = "x-builtin-port-versions-dir";
+        std::unique_ptr<std::string> builtin_port_versions_dir;
```
I think we should do a full bikeshed on these in the new year. I think they're still fine to bikeshed.

ras0219 reviewed (commented):
@strega-nil Please consider opening a new PR that addresses these issues
